### ISSUE TITLE
New Kernel v6.0.5-200 Breaks Audio

### ISSUE THREAD

**Issue by huang-jy** [jmontleon/pixelbook-fedora#47]
Just reporting this as it appears the new 6.x kernel has dropped and sound is broken on that. The Speakers aren't showing.

Booting off previous kernel (5.19.16-200) -- sound works, so it's definitely something in (or not in) the new kernel.

**Comment by admiral0** [comment]
Can confirm - both on Fedora and NixOS

**Comment by jmontleon** [comment]
Worth filing a bug on the kernel bugzilla if you have a minute. If not, I'll get to it when I can. There's no custom kernel at this point (not to say there can't be if we need a temporary patch), so there's nothing reporting here will do to help until we know more.

**Comment by jmontleon** [comment]
https://bugzilla.kernel.org/show_bug.cgi?id=216641

**Comment by jmontleon** [comment]
I don't see anything telling yet. Probably time to look at commits around this code.

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/sound/soc/intel/skylake?h=v6.0.6

Probably easy enough to revert one by one and find the culprit. Could be something else out of left field, but my money is on one of the top 5 there since 5.18.

**Comment by jmontleon** [comment]
It's been broken at least since [kernel-6.0.0-0.rc1.13.fc38](https://koji.fedoraproject.org/koji/buildinfo?buildID=2046012)

**Comment by jmontleon** [comment]
[5.19.17 works so it's ](https://koji.fedoraproject.org/koji/buildinfo?buildID=2080019) works, so that's probably about as narrowed down as we're getting without someone who knows more or a bisect.

**Comment by jmontleon** [comment]
I have to check myself to believe it... ya about 14k commits in 6.0 rc1. Gonna be a fun one to figure out if no one has an idea upstream. I fired off a build to split the first ~7000.

Venting, but I'm honestly ready to give up on this hardware; every other kernel update horribly breaks something and it becomes a game of bisect hundreds or thousands of commits to find out who/what broke it.

**Comment by jmontleon** [comment]
https://www.linuxcompatible.org/story/linux-kernel-60rc1-released/

**Comment by jmontleon** [comment]
I've narrowed it down to one of about 54 commits, although it's really a lot better than that because ~45 of them are driver changes related to one core ASoC change. Maybe in the next 24-48 I can update the kernel bugzilla and provide a patched kernel.

**Comment by jmontleon** [comment]
Leaving my list of reverts. I guess squashing commits isn't a thing with the kernel. Like 45 of these can't live without each other; I don't know why they're separate.

```
cc7afe6c259e486cfbbef4be05f8482fc355ec31 test
00a3ab5fb2a0a6d8c6bc029333cf05730119a807 Revert "ASoC: mxs-saif: Update to use set_fmt_new callback"
08db5100ddb82c874a5e6303aa7e6f22c0f926f6 Revert "ASoC: rockchip: Update to use set_fmt_new callback"
68d1e093da69a5d392b66cf78a3cacc5c7ba2724 Revert "ASoC: qcom: Update to use set_fmt_new callback"
293bcce9660b67ba321728d432bcd5a47af2c731 Revert "ASoC: meson: Update to use set_fmt_new callback"
3c893fda50cacae29ed031b2f90aea681ae7b249 Revert "ASoC: mediatek: mt8186: Use new direct clock defines"
719958e7572047a864ee11b813cf689dfb5b5baf Revert "ASoC: mediatek: Update to use set_fmt_new callback"
be66623f6e89cc3381cfd690282f68c4d328e2e0 Revert "ASoC: img: Update to use set_fmt_new callback"
5a10466b53973e59ccac2dd8c64d9c7e1c31687b Revert "ASoC: hisilicon: Update to use set_fmt_new callback"
d980b123e9699dea4fe4db3df9e775debfc74a26 Revert "ASoC: test-component: Update to use set_fmt_new callback"
33c26404c7c7e89a49c887d2f3009780897c8c71 Revert "ASoC: samsung: Update to use set_fmt_new callback"
213652c3081ba179a946c1064601cbc897c6aec2 Revert "ASoC: sh: Update to use set_fmt_new callback"
dab7dee1b1509d044d444dc07a0afb569a239c08 Revert "ASoC: pxa: Update to use set_fmt_new callback"
b9311335db9500a2969f21dece7ff7f1a7c2140a Revert "ASoC: stm: Update to use set_fmt_new callback"
c632281516bac9fdcba9a48e98fd8368332a382f Revert "ASoC: tegra: Update to use set_fmt_new callback"
8daced6611324e11c84fb7aac75ca2cb861a84db Revert "ASoC: xtensa: Update to use set_fmt_new callback"
48257ed94c86abecddc1243eeca449cc01c57c8e Revert "ASoC: tegra: Fix clock DAI format on Tegra210"
49353dcbd027f6c7102beba05e14d6338f972fbc Revert "ASoC: ti: Update to use set_fmt_new callback"
21074e8aab31c8ee05ed04e682401e381b3e44b8 Revert "ASoC: sunxi: Update to use set_fmt_new callback"
30bcff739ad362b315d0f4af60189929114edd10 Revert "ASoC: fsl: Update to use set_fmt_new callback"
2e518b53d3f52d520d84493d5b2ce9ac46c5d50c Revert "ASoC: ux500: Update to use set_fmt_new callback"
ec2044769c830db37ec5c0f27adac66133606ba6 Revert "ASoC: Intel: Update to use set_fmt_new callback"
8e06645831c069af76c73a31b0cdbd93b16af0d0 Revert "ASoC: js4740-i2s: Update to use set_fmt_new callback"
00351478257a908b151c5592f8ba99c8e35af4f0 Revert "ASoC: atmel: Update to use set_fmt_new callback"
34ebd4e0e9a58245b14d904fec0d6f0effe64133 Revert "ASoC: au1x: Update to use set_fmt_new callback"
805061f92fc0283c8e2a671fcda7b1512a17d0b0 Revert "ASoC: bcm: Update to use set_fmt_new callback"
417615244ad9b762ceac4a78cc695f33f53b570e Revert "ASoC: ep93xx: Update to use set_fmt_new callback"
32832a8152c638d3c4212dc3e202855ee2ed4c1f Revert "ASoC: dwc: Update to use set_fmt_new callback"
403d8e55f5a47f3f66d4dbac55ebab51cc2fefab Revert "ASoC: mediatek: Rename set_fmt_new back to set_fmt"
c93e86e0979e7c67d6b74dba89649b3f7a5661d6 Revert "ASoC: meson: Rename set_fmt_new back to set_fmt"
8964315518484394f55ecf2ca88c1d11d0451e37 Revert "ASoC: mxs-saif: Rename set_fmt_new back to set_fmt"
ebe890af019a2d5fad4e3bccb554f52729b34735 Revert "ASoC: pxa: Rename set_fmt_new back to set_fmt"
dd2e642d2c2f3c08aa60c9c242348c1a6f6988d9 Revert "ASoC: qcom: Rename set_fmt_new back to set_fmt"
3ab68b4757336a2d0166418e95aa13cbdec2ae6f Revert "ASoC: rockchip: Rename set_fmt_new back to set_fmt"
7bf51fdf6954c26316a36a2b13b6d3d50afb5f02 Revert "ASoC: samsung: Rename set_fmt_new back to set_fmt"
7546be42139d021fc708884cd0d62a9648b270c1 Revert "ASoC: sh: Rename set_fmt_new back to set_fmt"
c176702be4d2410456eea84eb78a860a26acd79e Revert "ASoC: stm: Rename set_fmt_new back to set_fmt"
e16e6b3a8c8c993d0449a07cc54cd605362aa0ea Revert "ASoC: sunxi: Rename set_fmt_new back to set_fmt"
64bdd9654b6a8d13f22869fcdf4fd7b5e216dea1 Revert "ASoC: tegra: Rename set_fmt_new back to set_fmt"
43f5168fc6dcee5b3099db28bc0e18194014ce47 Revert "ASoC: test-component: Rename set_fmt_new back to set_fmt"
619816f42a4e77ee12758504dc6d647dcb7da69a Revert "ASoC: ti: Rename set_fmt_new back to set_fmt"
7dcd8033e50ecdb28d01cac030221c0c4c9db87a Revert "ASoC: ux500: Rename set_fmt_new back to set_fmt"
d33266131c902e0dcc037c501ecd2416f4c008c3 Revert "ASoC: xtensa: Rename set_fmt_new back to set_fmt"
3e0308b4f5929d1ccead50d319c18ce82782a8dc Revert "ASoC: amd: vangogh: Update to use set_fmt_new callback"
6161b94257ba3ca657016bf51c9a76cbacb17925 Revert "ASoC: soc-core: remove always-false path"
25878d1201aa92bf89f1772d45798ca4ddbbd601 Revert "ASoC: core: Correct spelling fliped -> flipped"
5317b7cd196396a88d9aa1346baf1b2afd2689c7 Revert "ASoC: core: Pass legacy_dai_naming flag directly"
8721d06d6902c18a6c17b5f42d5dc269a0c32d54 Revert "ASoC: core: Add set_fmt_new callback that directly specifies provider"
307ca95f48a2fd65c6aef4bec77b125a294b4e1b Revert "ASoC: core: Always send the CPU DAI a direct clock specifier"
922ec1ffb0fcb71a8a0d65157f1622f1038169c4 Revert "ASoC: Introduce 'fixup_controls' card method"
6226d9cc152df54464fa08a3ddd978098dc1666e Revert "ASoC: core: Make snd_soc_unregister_card() return void"
378a4c5ae2fa379bb7a6dca20512811ba8013d34 Revert "ASoC: core: Switch core to new DAI naming flag"
d6c16a3ab98e28a500bd1d4954599449188a5216 Revert "ASoC: soc-core.c: fixup snd_soc_of_get_dai_link_cpus()"
6302b1ce9d50d08acf452cfa39bf8e1323418aee Revert "ASoC: soc-core.c: share code for snd_soc_of_get_dai_link_cpus/codecs()"
23fb1d94950e513693460c17f27a8722a5514045 Revert "ASoC: Intel: skylake: skl-pcm: use pm_runtime_resume_and_get()"
3829606fc5dffeccdf80aebeed3aa75255257f35 Linux 6.0.5
```

**Comment by jmontleon** [comment]
These are the interesting ones. The rest I believe are related to 8721d06d6902c18a6c17b5f42d5dc269a0c32d54 if I remember correctly

```
6161b94257ba3ca657016bf51c9a76cbacb17925 Revert "ASoC: soc-core: remove always-false path"
25878d1201aa92bf89f1772d45798ca4ddbbd601 Revert "ASoC: core: Correct spelling fliped -> flipped"
5317b7cd196396a88d9aa1346baf1b2afd2689c7 Revert "ASoC: core: Pass legacy_dai_naming flag directly"
8721d06d6902c18a6c17b5f42d5dc269a0c32d54 Revert "ASoC: core: Add set_fmt_new callback that directly specifies provider"
307ca95f48a2fd65c6aef4bec77b125a294b4e1b Revert "ASoC: core: Always send the CPU DAI a direct clock specifier"
922ec1ffb0fcb71a8a0d65157f1622f1038169c4 Revert "ASoC: Introduce 'fixup_controls' card method"
6226d9cc152df54464fa08a3ddd978098dc1666e Revert "ASoC: core: Make snd_soc_unregister_card() return void"
378a4c5ae2fa379bb7a6dca20512811ba8013d34 Revert "ASoC: core: Switch core to new DAI naming flag"
d6c16a3ab98e28a500bd1d4954599449188a5216 Revert "ASoC: soc-core.c: fixup snd_soc_of_get_dai_link_cpus()"
6302b1ce9d50d08acf452cfa39bf8e1323418aee Revert "ASoC: soc-core.c: share code for snd_soc_of_get_dai_link_cpus/codecs()"
23fb1d94950e513693460c17f27a8722a5514045 Revert "ASoC: Intel: skylake: skl-pcm: use pm_runtime_resume_and_get()"
```

**Comment by jmontleon** [comment]
One of these is the culprit
```
586fb2641371cf7f23a401ab1c79b17e3ec457f4 "ASoC: soc-core.c: fixup snd_soc_of_get_dai_link_cpus()"
1892a991886ace2c3450bec801df2cf4028a803a "ASoC: core: Make snd_soc_unregister_card() return void"
df4d27b19b892f464685ea45fa6132dd1a2b6864 "ASoC: Introduce 'fixup_controls' card method"
6c076273a326cc5b5162451aacf7b7744bb03c66 "ASoC: core: Always send the CPU DAI a direct clock specifier"
8c8a0f01c7c52f9037b6859ff5234ea5acf572d6 "ASoC: core: Pass legacy_dai_naming flag directly"
8ad0b83e0079564989ccbe5b864cc265a0f5a14d "ASoC: soc-core: remove always-false path"
```

**Comment by jmontleon** [comment]
These commits and more are an unbelievable knot. Even when possible to cleanly revert them builds routinely fail because of other interdependent PRs.

If someone from upstream doesn't respond with a fix in the next few days I'll be archiving the repo and looking for new hardware. I'm tired of regressions, particularly with the sound system. This is exactly what is NOT supposed to happen.

https://docs.kernel.org/process/handling-regressions.html

And yet this, the recent wifi/tb breakage, _at least_ one prior audo regression from the kernel, not to menton alsa ucm changes breaking it at least twice.

**Comment by huang-jy** [comment]
I am sorry you're hitting all these problems. I understand the frustrations and why you're considering archiving. With the recent spate of breakages affecting this hardware it is indeed very annoying, and I'm sure you're not the only one (most likely breaking one hardware will break others in a similar way)

**Comment by admiral0** [comment]
+1 this hardware is incredibly annoying....

**Comment by jmontleon** [comment]
I have it down to:
```
ff31753fcb061b90bd8c356d5b27a6eb5f8ade15 "ASoC: simple-card-utils: rename asoc_simple_init_dai_link_params() to asoc_simple_init_for_codec2codec()"
28086d05ada6d03daa886aad0e469854b811311c "ASoC: simple-card-utils: Move snd_soc_component_is_codec to be local"
129f055a2144ab588a43c2e66d21a1f55ce54f81 "ASoC: core: Switch core to new DAI naming flag"
8c8a0f01c7c52f9037b6859ff5234ea5acf572d6 "ASoC: core: Pass legacy_dai_naming flag directly"
```

I am doing builds trying to find out what one or combo is the culprit from here.

I sent an email to hoping to get some visibility / help with the constant regressions, but I don't know how much if anything we can expect. 

If anyone wants 6.0.5 with working sound there's a build going:
https://copr.fedorainfracloud.org/coprs/jmontleon/pixelbook/build/5001262/

That has about 11 commits reverted; I've been working backwards and when I either find the one/combo or get stuck tonight I'll fire off another build with fewer commits removed. It should be possible to patch out the problem in the very short term.

**Comment by jmontleon** [comment]
I have a one line fix and will submit a patch to the kernel. An updated kernel is building.

**Comment by jmontleon** [comment]
I take it back. I've narrowed it down to a few patches. I don't know what I just thought I did, but it wasn't test my last build. I'm pretty sure it's going to come down to a one line change; I'm just trying to dissect where/what is missing the setting.

**Comment by jmontleon** [comment]
If anyone is interested the full set of reverts look like this; it's a lot of renames and not much else:

Coming from reverting these commits:

A lot of this is renaming / moving a function to make it private. 

28086d05ada6d03daa886aad0e469854b811311c
129f055a2144ab588a43c2e66d21a1f55ce54f81
8c8a0f01c7c52f9037b6859ff5234ea5acf572d6

```
diff --git a/include/sound/soc-component.h b/include/sound/soc-component.h
index c26ffb033777..feec3efd0a82 100644
--- a/include/sound/soc-component.h
+++ b/include/sound/soc-component.h
@@ -348,6 +348,11 @@ static inline int snd_soc_component_cache_sync(
        return regcache_sync(component->regmap);
 }
 
+static inline int snd_soc_component_is_codec(struct snd_soc_component *component)
+{
+       return component->driver->legacy_dai_naming;
+}
+
 void snd_soc_component_set_aux(struct snd_soc_component *component,
                               struct snd_soc_aux_dev *aux);
 int snd_soc_component_init(struct snd_soc_component *component);
diff --git a/sound/soc/generic/simple-card-utils.c b/sound/soc/generic/simple-card-utils.c
index ef08d6bb57ba..d5907412ff09 100644
--- a/sound/soc/generic/simple-card-utils.c
+++ b/sound/soc/generic/simple-card-utils.c
@@ -513,11 +513,6 @@ static int asoc_simple_init_dai(struct snd_soc_dai *dai,
        return 0;
 }
 
-static inline int asoc_simple_component_is_codec(struct snd_soc_component *component)
-{
-       return component->driver->endianness;
-}
-
 static int asoc_simple_init_dai_link_params(struct snd_soc_pcm_runtime *rtd,
                                            struct simple_dai_props *dai_props)
 {
@@ -537,7 +532,7 @@ static int asoc_simple_init_dai_link_params(struct snd_soc_pcm_runtime *rtd,
 
        /* Only Codecs */
        for_each_rtd_components(rtd, i, component) {
-               if (!asoc_simple_component_is_codec(component))
+               if (!snd_soc_component_is_codec(component))
                        return 0;
        }
 
diff --git a/sound/soc/intel/boards/kbl_rt5663_rt5514_max98927.c b/sound/soc/intel/boards/kbl_rt5663_rt5514_max98927.c
index 2c79fca57b19..85a0d6aba1a5 100644
--- a/sound/soc/intel/boards/kbl_rt5663_rt5514_max98927.c
+++ b/sound/soc/intel/boards/kbl_rt5663_rt5514_max98927.c
@@ -856,6 +856,7 @@ static struct platform_driver kabylake_audio = {
        .driver = {
                .name = "kbl_r5514_5663_max",
                .pm = &snd_soc_pm_ops,
+               .legacy_dai_naming = 1,
        },
        .id_table = kbl_board_ids,
 };
diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index e824ff1a9fc0..0679f3567326 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -2488,7 +2488,7 @@ static int snd_soc_register_dais(struct snd_soc_component *component,
 
        for (i = 0; i < count; i++) {
                dai = snd_soc_register_dai(component, dai_drv + i, count == 1 &&
-                                          component->driver->legacy_dai_naming);
+                                          !snd_soc_component_is_codec(component));
                if (dai == NULL) {
                        ret = -ENOMEM;
                        goto err;

```

I think this revert is what's fixing it, and when we go in reverse it breaks.
```
-                                          component->driver->legacy_dai_naming);
+                                          !snd_soc_component_is_codec(component));
```

If I track we're getting the opposite of `snd_soc_component_is_codec` when it works, which is just returning `return component->driver->legacy_dai_naming;`. Yet after teh switch we're getting. `component->driver->legacy_dai_naming)`.

Part of this mess is that `non_legacy_dai_naming` was inverted to `legacy_dai_naming` and when I revert 28086d05ada6d03daa886aad0e469854b811311c I have to edit/flip ` return component->driver->non_legacy_dai_naming;` to ` return component->driver->legacy_dai_naming;` or the build fails because `non_legacy_dai_naming` no is no longer defined. 

I think what's happening, and I say this with a biig caveat that I have no freaking idea what I'm looking at, is that legacy_dai_naming just needs to be set to 1 somewhere.

Whenever the module is loaded we just get stuck:
```
[   16.928454] kbl_r5514_5663_max kbl_r5514_5663_max: ASoC: CPU DAI spi-PRP0001:00 not registered
[   16.928561] platform kbl_r5514_5663_max: deferred probe pending
```

spi-PRP0001:00 is only found at:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/sound/soc/intel/boards/kbl_rt5663_rt5514_max98927.c?h=v6.0.6#n527

Corresponding with the rt5514 dsp down below I believe:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/sound/soc/intel/boards/kbl_rt5663_rt5514_max98927.c?h=v6.0.6#n609

So maybe the 5514 driver or something. I could be wholly and completely off the mark, but it seems to be around here we're losing it.

**Comment by jmontleon** [comment]
Cementing my feeling that we have an inverted value somewhere; sound works with this one line change:

```
diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index e824ff1a9fc0..f4a11f8f53bf 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -2488,7 +2488,7 @@ static int snd_soc_register_dais(struct snd_soc_component *component,
 
        for (i = 0; i < count; i++) {
                dai = snd_soc_register_dai(component, dai_drv + i, count == 1 &&
-                                          component->driver->legacy_dai_naming);
+                                          !component->driver->legacy_dai_naming);
                if (dai == NULL) {
                        ret = -ENOMEM;
                        goto err;

```

**Comment by jmontleon** [comment]
Seeing if I can add some debug  to find the driver name and the value; if I can get that it should hopefully be a straight shot.

**Comment by jmontleon** [comment]
`rt5514 spi-PRP0001:00: component name rt5514-spi legacy_dai_naming value: 1` and now it works. I'll build a kernel and submit a PR.

**Comment by jmontleon** [comment]
Looks like 6.0.6 is up in koji so building a new one. Should finish in the coming hours:
https://copr.fedorainfracloud.org/coprs/jmontleon/pixelbook/build/5004382/

See if the PR to mainliune just gets me yelled at, a different fix is suggested, or flat out ignored.
https://lore.kernel.org/regressions/20221102200527.4174076-1-jmontleo@redhat.com/T/#u

**Comment by jmontleon** [comment]
Looks like I'll have to send off a new version to fix the commit message at the least. When some more needed feedback comes in I'll send out the update and hopefully we can get it merged.

**Comment by jmontleon** [comment]
On the right track I guess. ACK'd after adding the fix for another driver and some other go round for tags/signoff.
https://lore.kernel.org/regressions/20221103150700.GC10437@ediswmail.ad.cirrus.com/T/#u

**Comment by jmontleon** [comment]
As a side note I came across this note when digging through lots of stuff the other night:

https://www.alsa-project.org/wiki/ASoC

There is no standard method to signal user initiated audio events. e.g. Headphone/Mic insertion, Headphone/Mic detection after an insertion event. These are quite common events on portable devices and often require machine specific code to re route audio, enable amps etc after such an event.

Which I guess explains why aside from the workaround we have this has never worked.

**Comment by jmontleon** [comment]
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?h=next-20221107&id=392cc13c5ec72ccd6bbfb1bc2339502cc59dd285

**Comment by jmontleon** [comment]
6.1-rc6 has the fix https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/sound/soc/codecs?h=v6.1-rc6
It's in the stable queue for 6.0 as well. https://git.kernel.org/pub/scm/linux/kernel/git/stable/stable-queue.git/tree/queue-6.0

**Comment by jmontleon** [comment]
We should be back to the stock kernel with 6.0.10.

**Comment by jmontleon** [comment]
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/?h=v6.0.10&qt=author&q=montleon

**Comment by jmontleon** [comment]
6.0.10 is now available with the fix in stable for Fedora 35, 36, and 37:
https://bodhi.fedoraproject.org/updates/FEDORA-2022-78f2005ce4
https://bodhi.fedoraproject.org/updates/FEDORA-2022-48b826559f
https://bodhi.fedoraproject.org/updates/FEDORA-2022-a40f54d6fc

**Comment by gabrys** [comment]
@jmontleon, I wanted to jump in and say, thank you SO MUCH for supporting this effort. I really like the form factor and the looks of the Pixelbook (just not the ChromeOS) and I'm it daily as my primary Linux machine, thanks to your work.

I found this issue when I started searching for whether Pixelbook is supported on "your" Fedora after upgrading to 37 and I have seen so much effort put into fixing the issue blocking it from fully working.

If you have a way to support you financially as a way to say THANK$$$, let me know, I'll send a donation.
